experiments/process_sandbox/include/process_sandbox/netpolicy.h

```diff
@@ -0,0 +1,245 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+/**
+ * This file contains the abstractions required for granting network access in
+ * a sandbox.  Sandboxes are not permitted to do anything directly that touches
+ * a global namespace and so may not connect or bind sockets.  Any attempt to
+ * use the network must be proxied.
+ */
+
+#pragma once
+#include <netdb.h>
+#include <optional>
+#include <string>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <variant>
+
+namespace sandbox
+{
+  /**
+   * Class encapsulating a network access policy for a sandbox.  This allows
+   * specific access to some network resources to be granted to a sandbox.
+   */
+  class NetworkPolicy
+  {
+    /**
+     * C++17 made noexcept part of the type system, so `R(Args...) noexcept` is
+     * not the same as `R(Args...)`.  Unfortunately, `std::function` remains
+     * specialised only for `R(Args...)` and so you cannot create a
+     * `std::function<R(Args...) noexcept>`.  Fortunately, the `noexcept`
+     * version is a subtype of the potentially throwing version (throwing 0
+     * exceptions is a stronger constraint than throwing 0 or more exceptions)
+     * and so we can just strip `noexcept` from the types.
```

Comment:
I'd expect the type system to do that for you. Is there a defect reported on this?

Reply:
I posted a comment about this in the MSR Cambridge Engineering teams channel, but here's a summary:

Prior to C++17, `noexcept` was not part of the type system.  It is now, but `R(Args...) noexcept` is a subtype of `R(Args...)` because throwing zero exceptions is a strictly weaker constraint than throwing zero or more exceptions.  The call method on `std::function` is allowed to throw if on `std::function<R(Args...)>(nullptr)`, so capturing whether the wrapped function is `noexcept` or not is not of value to `std::function` (the caller must assume that it can throw in any situation).  As a result, `std::function<R(Args...) noexcept>` would have to behave exactly as `std::function<R(Args...)>` and so there is no benefit in allowing the first specialisation.  Because the argument of the first is a subtype of the argument of the second, we can always just implicitly cast away the `noexcept` when creating a `std::function`.    As a result, this is not viewed as a defect in the C++17 standard.

This is only a problem when the type of the `std::function` is deduced from the initialiser.  In this case, the deduced type contains `noexcept` and (at least with libstdc++) the instantiation of `std::function` fails.  

I think the only real defect here is that there isn't a `std::remove_noexcept` in `<type_traits>`.